### PR TITLE
Static tests: forbid 'or' instead of '||' #21062.

### PR DESCRIPTION
**Pull request by novikor**
### Description (*)
Added `Squiz.Operators.ValidLogicalOperators` PHP-CS rule

### Fixed Issues (if relevant)

1. magento/magento2#21062: Static tests: forbid 'or' instead of '||' 

### Manual testing scenarios (*)
Just make sure PHP-CS asks you to avoid the literal logical operators
image
### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @novikor. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by sivaschenko** [comment]
Hi @novikor, thanks for the pull request! Considering that `logical_operators` is a risky fixer (more details https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/1212), I wouldn't add it to the default list of fixers. I think it's better to allow the individuals who are confident about it to add the fixer to their local configuration file.

image

**Review by orlangur — CHANGES_REQUESTED** [review]
Having risky rule is totally fine while we have static test which uses `php-cs-fixer` in dry run mode.

Please add such test and also corresponding `phpcs` sniff wouldn't hurt.

Please preserve alphabetical order of fixers.

**Comment by novikor** [comment]
> Having risky rule is totally fine while we have static test which uses `php-cs-fixer` in dry run mode.
> 
> Please add such test and also corresponding `phpcs` sniff wouldn't hurt.
> 
> Please preserve alphabetical order of fixers.

Hi.
I have already added this rule into static tests and phpcs: https://github.com/magento/magento-coding-standard/pull/40
As I understand, nothing can be done anymore in this PR?

**Comment by novikor** [comment]
@orlangur , may I just update phpunit.xml.dist from https://github.com/magento/magento-coding-standard ?

**Comment by orlangur** [comment]
@novikor `php-cs-fixer` integration could be done in this PR :)

What would you like to add from `phpunit.xml.dist`?

**Comment by novikor** [comment]
@orlangur , I am a little bit confused about `php-cs-fixer`. As it was mentioned before, it is risky rule so it is not recommended to add into default rules list.

I would like to add changes from https://github.com/magento/magento-coding-standard/pull/40/files#diff-eb0df15a533f0e9054dc4d489ee5e67a

**Comment by orlangur** [comment]
@novikor mentioned changes are ok to be added.

Risky rule will be ok if a test like https://github.com/magento/magento2/blob/2.3-develop/dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php#L262 is implemented for `php-cs-fixer`.

**Comment by sivaschenko** [comment]
@orlangur @novikor actually I am concerned about forcing developers to use "allow-risky" option in order to be able to run cs fixer using the default profile.

**Comment by orlangur** [comment]
@sivaschenko ok, agree, although relying on operator precedence does not look like a realistic scenario such developer experience would be pretty confusing. We may use risky fixers for dry run mode only when we have a `php-cs-fixer` test.

@novikor feel free to contribute `phpcs` rule only and `php-cs-fixer` test may be contributed as a separate improvement.

**Comment by novikor** [comment]
@orlangur , I decided to improve `phpcs` only. Please review.
image

File: dev/tests/static/framework/Magento/ruleset.xml [hunk 1]
```diff
@@ -85,4 +85,8 @@
     <rule ref="Squiz.Commenting.DocCommentAlignment"/>
     <rule ref="Squiz.Functions.GlobalFunction"/>
     <rule ref="Squiz.WhiteSpace.LogicalOperatorSpacing"/>
+    <rule ref="Squiz.Operators.ValidLogicalOperators">
+        <severity>6</severity>
```
Comment (orlangur) [inline dev/tests/static/framework/Magento/ruleset.xml]: No need to spoil core coding standard with severities, each rule is mandatory.

Please put it before previous line to preserve alphabetical order.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: dev/tests/static/framework/Magento/ruleset.xml (hunk 1 above)
Comment (novikor) [reply dev/tests/static/framework/Magento/ruleset.xml]: Done

File: dev/tests/static/framework/Magento/ruleset.xml [hunk 2]
```diff
@@ -84,5 +84,8 @@
 
     <rule ref="Squiz.Commenting.DocCommentAlignment"/>
     <rule ref="Squiz.Functions.GlobalFunction"/>
+    <rule ref="Squiz.Operators.ValidLogicalOperators">
+        <type>warning</type>
```
Comment (orlangur) [inline dev/tests/static/framework/Magento/ruleset.xml]: No need to make it warning, any rule is mandatory.

**Review by orlangur — CHANGES_REQUESTED** [review]
No need to make it warning, any rule is mandatory. Please squash into one commit.

**Comment by novikor** [comment]
> No need to make it warning, any rule is mandatory. Please squash into one commit.

Done)

**Review by orlangur — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-4300](https://jira.corp.magento.com/browse/ENGCOM-4300) has been created to process this Pull Request

**Pull request merged**